compositor_api/src/types/from_color.rs

```diff
@@ -0,0 +1,268 @@
+use compositor_render::scene;
+
+use super::util::*;
+
+impl TryFrom<RGBAColor> for scene::RGBAColor {
+    type Error = TypeError;
+    fn try_from(value: RGBAColor) -> std::result::Result<Self, Self::Error> {
+        let s = &value.0.trim();
+        match s.chars().next() {
+            Some('#') => match s.len() {
+                7 => parse_hex(s),
+                9 => parse_hex_rgba(s),
+                _ => Err(TypeError::new(
+                    "Invalid format. Color has to be in #RRGGBB or #RRGGBBAA format.",
+                )),
+            },
+            Some('r') => {
+                if s.starts_with("rgb(") {
+                    parse_rgb(s)
+                } else if s.starts_with("rgba(") {
+                    parse_rgba(s)
+                } else {
+                    Err(TypeError::new(
+                        "Invalid format. Expected rgb() or rgba() format.",
+                    ))
+                }
+            }
+            _ => {
+                if let Some(rgb_str) = parse_named_color(s) {
+                    parse_rgb(rgb_str)
+                } else {
+                    Err(TypeError::new("Unsupported color format."))
+                }
+            }
+        }
+    }
+}
+
+fn parse_hex(s: &str) -> Result<scene::RGBAColor, TypeError> {
+    let (r, g, b) = (&s[1..3], &s[3..5], &s[5..7]);
+    let a = "ff"; // default full opacity
+    parse_color_components(r, g, b, a)
+}
+
+fn parse_hex_rgba(s: &str) -> Result<scene::RGBAColor, TypeError> {
+    let (r, g, b, a) = (&s[1..3], &s[3..5], &s[5..7], &s[7..9]);
+    parse_color_components(r, g, b, a)
+}
+
+fn parse_rgb(s: &str) -> Result<scene::RGBAColor, TypeError> {
+    let inner_s = s.trim_start_matches("rgb(").trim_end_matches(')');
+    let parts: Vec<&str> = inner_s.split(',').collect();
+    if parts.len() != 3 {
+        return Err(TypeError::new("Invalid RGB format."));
+    }
+    let (r, g, b) = (parts[0].trim(), parts[1].trim(), parts[2].trim());
+    let a = "255"; // default full opacity
+    parse_color_components_from_decimal(r, g, b, a)
+}
+
+fn parse_rgba(s: &str) -> Result<scene::RGBAColor, TypeError> {
+    let inner_s = s.trim_start_matches("rgba(").trim_end_matches(')');
+    let parts: Vec<&str> = inner_s.split(',').collect();
```

Comment:
You could do sth like this
```suggestion
    let parts: Vec<u32> = inner_s.split(',').map(|c|parse_color_channel(c.trim())).collect::<Result<Vec<_>,_>()?;
```
not sure if the `::<Result<Vec<_>,_>` will be needed or Rust can figure out the type

Reply:
it couldn't be handled like that because alpha in the original format is a float between 0 and 1. I've processed it slightly differently as a whole